### PR TITLE
bug fix, performance improvement, new features

### PR DESCRIPTION
**Pull request by snipking**
## Purpose
1. use long type sequence
2. update example user-mgt example
3. support unset attribute
4. support query limit result count
5. add query test class to help debug
6. multiple query bug fix
7. support $or query statment
8. support mobile as username
9. case insensitive username query performance optimize
10. fix getIncrementedSequence concurrent issue
11. fix typo

## Goals
As Purpose described.

## Approach
N/A

## User stories
1. login with user mobile number
2. list users more than 1000
3. increace case insensitive username query performance

## Release note
1. use long type sequence
2. update example user-mgt example
3. support unset attribute
4. support query limit result count
5. add query test class to help debug
6. multiple query bug fix
7. support $or query statment
8. support mobile as username
9. case insensitive username query performance optimize
10. fix getIncrementedSequence concurrent issue
11. fix typo

## Documentation
N/A

## Training
N/A

## Certification
N/A

## Marketing
N/A

## Automation tests
N/A

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? yes
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

## Samples
N/A

## Related PRs
N/A

## Migrations (if applicable)
1. collection `UM_USER` add attribute `UM_CASE_INSENSITIVE_USER_NAME` which contains uppercase username.
2. collection `UM_USER_ATTRIBUTE` add attribute `caseInsensitiveUid` which contains uppercase uid.

## Test environment
Oracle JDK 1.8.0_192-b12
OS: macos / Alpine
Database: mysql 5.7 / mongodb 3.4.7
Browser: chrome 71
 
## Learning
N/A

### CODE REVIEW

File: src/main/java/org/wso2/carbon/mongodb/query/MongoPreparedStatementImpl.java [hunk 1]
```diff
@@ -491,6 +547,38 @@ private void getUpdateObject(JSONObject object) {
         this.query = new BasicDBObject(mapQuery);
     }
 
+    /**
+     * get the unset object.
+     *
+     * @param object to update
+     */
+    private void getUnsetObject(JSONObject object) {
```
Comment (ruwanta) [inline src/main/java/org/wso2/carbon/mongodb/query/MongoPreparedStatementImpl.java:555]: Returning void on getter method looks odd. There are some other similar wrong pattern on the existing code in this class. Yet new code need to be done properly.

File: src/main/java/org/wso2/carbon/mongodb/util/MongoDatabaseUtil.java [hunk 2]
```diff
@@ -78,6 +75,11 @@ public static synchronized DB getRealmDataSource(RealmConfiguration realmConfigu
      * @throws UserStoreException if any error occurred
      */
     public static DB createRealmDataSource(RealmConfiguration realmConfiguration) throws UserStoreException {
+
+        if (log.isDebugEnabled()) {
+            log.debug("createRealmDataSource(RealmConfiguration realmConfiguration)");
```
Comment (ruwanta) [inline src/main/java/org/wso2/carbon/mongodb/util/MongoDatabaseUtil.java:80]: can we change this to "Creating mongo datasource with realm configuration "+ realmConfiguration+getDescription

File: src/main/java/org/wso2/carbon/mongodb/util/MongoDatabaseUtil.java [hunk 3]
```diff
@@ -63,6 +55,11 @@
      * @throws UserStoreException if any error occurred
      */
     public static synchronized DB getRealmDataSource(RealmConfiguration realmConfiguration) throws UserStoreException {
+
+        if (log.isDebugEnabled()) {
+            log.debug("getRealmDataSource(RealmConfiguration realmConfiguration)");
```
Comment (ruwanta) [inline src/main/java/org/wso2/carbon/mongodb/util/MongoDatabaseUtil.java:60]: can we change this to "Creating mongo datasource with realm configuration "+ realmConfiguration+getDescription

File: src/main/java/org/wso2/carbon/mongodb/util/MongoDatabaseUtil.java [hunk 4]
```diff
@@ -117,6 +117,27 @@ public static DB createRealmDataSource(RealmConfiguration realmConfiguration) th
...
+                mongoClients = new HashMap<>();
+            } else {
+                mongoClient = mongoClients.get(urlWithCredentials);
+            }
+            if (mongoClient == null) {
+                mongoClient = new MongoClient(clientURI);
+                mongoClients.put(urlWithCredentials, mongoClient);
+
+                if(log.isDebugEnabled()) {
+                    log.debug("create new mongoClient");
```
Comment (ruwanta) [inline src/main/java/org/wso2/carbon/mongodb/util/MongoDatabaseUtil.java:132]: better change the log message "Created new mongoClient with username: " +userName

File: src/main/java/org/wso2/carbon/mongodb/util/MongoDatabaseUtil.java [hunk 5]
```diff
@@ -117,6 +117,27 @@ public static DB createRealmDataSource(RealmConfiguration realmConfiguration) th
...
+                }
+            } else {
+                if(log.isDebugEnabled()) {
+                    log.debug("use exist mongoClient");
```
Comment (ruwanta) [inline src/main/java/org/wso2/carbon/mongodb/util/MongoDatabaseUtil.java:136]: better change the log message "The user exists in the mongoClient with username: " +userName

**Review by ruwanta — COMMENTED** [review]
Few comments

**Comment by madurangasiriwardena** [comment]
@snipking,

Closing the PR, since the comments were not addressed for a long time. Please address the comments and re-open the PR.

**Pull request closed without merging**